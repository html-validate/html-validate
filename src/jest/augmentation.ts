/** module augmentation: jest.d.ts */
import { type ConfigData } from "../config";
import { type Message } from "../message";

declare global {
	/* eslint-disable-next-line @typescript-eslint/no-namespace -- module augmentation */
	namespace jest {
		/* eslint-disable-next-line @typescript-eslint/no-empty-object-type, @typescript-eslint/no-unused-vars -- to match jest declaration */
		interface Matchers<R, T = {}> {
			toBeValid(): R;
			toBeInvalid(): R;
			toHaveError(error: Partial<Message>): R;
			toHaveError(ruleId: string, message: string, context?: any): R;
			toHaveErrors(errors: Array<[string, string] | Record<string, unknown>>): R;

			/**
			 * Validate string or HTMLElement.
			 *
			 * Test passes if result is valid.
			 *
			 * @param config - Optional HTML-Validate configuration object.
			 * @param filename - Optional filename used when matching transformer and
			 * loading configuration.
			 */
			toHTMLValidate(filename?: string): R;
			toHTMLValidate(config: ConfigData, filename?: string): R;
			toHTMLValidate(error: Partial<Message>, filename?: string): R;
			toHTMLValidate(error: Partial<Message>, config: ConfigData, filename?: string): R;

			/**
			 * Writes out the given [[Report]] using codeframe formatter and compares
			 * with snapshot.
			 */
			toMatchCodeframe(snapshot?: string): R;

			/**
			 * Writes out the given [[Report]] using codeframe formatter and compares
			 * with inline snapshot.
			 */
			toMatchInlineCodeframe(snapshot?: string): R;
		}
	}
}
/** module augmentation end */
