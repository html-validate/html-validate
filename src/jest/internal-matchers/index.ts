export { default as toBeToken } from "./to-be-token";
