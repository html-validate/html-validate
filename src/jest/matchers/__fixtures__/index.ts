export { reportOk, reportOkAsync } from "./report-ok";
export { reportError, reportErrorAsync } from "./report-error";
export { reportMultipleErrors, reportMultipleErrorsAsync } from "./report-multiple-errors";
