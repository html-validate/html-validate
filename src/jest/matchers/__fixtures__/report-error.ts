import { Severity } from "../../../config";
import { Source } from "../../../context";
import { Report, Reporter } from "../../../reporter";

export function reportError(): Report {
	const reporter = new Reporter();
	const data = /* HTML */ `
		<header id="foo">
			<div>
				<p>lorem ipsum</p>
			</div>
		</header>
	`;
	const source: Source = {
		filename: "inline",
		line: 1,
		column: 1,
		offset: 0,
		data,
	};
	const offset = data.indexOf("lorem ipsum");
	reporter.addManual("inline", {
		ruleId: "my-rule",
		severity: Severity.ERROR,
		message: "mock message",
		line: 4,
		column: 8,
		offset,
		size: "lorem ipsum".length,
		selector: () => "#foo > div",
		context: {
			foo: "bar",
		},
	});
	return reporter.save([source]);
}

export function reportErrorAsync(): Promise<Report> {
	return Promise.resolve(reportError());
}
