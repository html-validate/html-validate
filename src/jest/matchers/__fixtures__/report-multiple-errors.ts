import { Severity } from "../../../config";
import { Report, Reporter } from "../../../reporter";

export function reportMultipleErrors(): Report {
	const reporter = new Reporter();
	reporter.addManual("inline", {
		ruleId: "my-rule",
		severity: Severity.ERROR,
		message: "mock message",
		line: 2,
		column: 15,
		offset: 43,
		size: 12,
		selector: null,
		context: {
			foo: "bar",
		},
	});
	reporter.addManual("inline", {
		ruleId: "another-rule",
		severity: Severity.ERROR,
		message: "another message",
		line: 2,
		column: 15,
		offset: 43,
		size: 12,
		selector: null,
	});
	return reporter.save();
}

export function reportMultipleErrorsAsync(): Promise<Report> {
	return Promise.resolve(reportMultipleErrors());
}
