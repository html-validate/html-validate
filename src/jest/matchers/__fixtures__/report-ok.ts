import { type Report, Reporter } from "../../../reporter";

export function reportOk(): Report {
	const reporter = new Reporter();
	return reporter.save();
}

export function reportOkAsync(): Promise<Report> {
	return Promise.resolve(reportOk());
}
