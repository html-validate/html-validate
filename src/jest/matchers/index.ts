export { toBeValid } from "./to-be-valid";
export { toBeInvalid } from "./to-be-invalid";
export { toHTMLValidate } from "./to-htmlvalidate";
export { toHaveError } from "./to-have-error";
export { toHaveErrors } from "./to-have-errors";
export { toMatchCodeframe } from "./to-match-codeframe";
export { toMatchInlineCodeframe } from "./to-match-inline-codeframe";
