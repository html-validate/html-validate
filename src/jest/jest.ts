import "./augmentation";

import {
	toBeValid,
	toBeInvalid,
	toHTMLValidate,
	toHaveError,
	toHaveErrors,
	toMatchCodeframe,
	toMatchInlineCodeframe,
} from "./matchers";
import { diff } from "./utils";

expect.extend({
	toBeValid: toBeValid(),
	toBeInvalid: toBeInvalid(),
	toHTMLValidate: toHTMLValidate(expect, diff),
	toHaveError: toHaveError(expect, diff),
	toHaveErrors: toHaveErrors(expect, diff),
	toMatchCodeframe: toMatchCodeframe(),
	toMatchInlineCodeframe: toMatchInlineCodeframe(),
});
