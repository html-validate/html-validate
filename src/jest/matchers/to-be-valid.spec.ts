import { reportError, reportErrorAsync, reportOk, reportOkAsync } from "./__fixtures__";
import "../jest";

describe("toBeValid()", () => {
	it("should pass if report is valid", () => {
		expect.assertions(1);
		expect(reportOk()).toBeValid();
	});

	it("should pass if async report is valid", async () => {
		expect.assertions(1);
		/* eslint-disable-next-line @typescript-eslint/await-thenable, @typescript-eslint/no-confusing-void-expression -- the types are wrong here, it does return a promise */
		await expect(reportOkAsync()).toBeValid();
	});

	it("should fail if report is invalid", () => {
		expect.assertions(3);
		let error: any;
		try {
			expect(reportError()).toBeValid();
		} catch (e: any) {
			error = e;
		}
		expect(error).toBeDefined();
		expect(error?.message).toMatchSnapshot();
	});

	it("should fail if async report is invalid", async () => {
		expect.assertions(3);
		let error: any;
		try {
			/* eslint-disable-next-line @typescript-eslint/await-thenable, @typescript-eslint/no-confusing-void-expression -- the types are wrong here, it does return a promise */
			await expect(reportErrorAsync()).toBeValid();
		} catch (e: any) {
			error = e;
		}
		expect(error).toBeDefined();
		expect(error?.message).toMatchSnapshot();
	});
});
