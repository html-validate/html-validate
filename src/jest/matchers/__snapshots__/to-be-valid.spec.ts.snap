// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`toBeValid() should fail if async report is invalid 1`] = `"Result should be valid but had error \\"mock message\\""`;

exports[`toBeValid() should fail if report is invalid 1`] = `"Result should be valid but had error \\"mock message\\""`;
